tests/Doctrine/Tests/ORM/Functional/Ticket/GH9335Test.php

```diff
@@ -0,0 +1,220 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\ORM\Functional\Ticket;
+
+use Doctrine\DBAL\ParameterType;
+use Doctrine\DBAL\Platforms\AbstractPlatform;
+use Doctrine\DBAL\Types\Type;
+use Doctrine\DBAL\Types\Type as DBALType;
+use Doctrine\ORM\Mapping\Column;
+use Doctrine\ORM\Mapping\Entity;
+use Doctrine\ORM\Mapping\Id;
+use Doctrine\ORM\Mapping\JoinColumn;
+use Doctrine\ORM\Mapping\OneToOne;
+use Doctrine\Tests\OrmFunctionalTestCase;
+
+/**
+ * @group GH9335
+ */
+final class GH9335Test extends OrmFunctionalTestCase
+{
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        if (! DBALType::hasType(GH9335IntObjectType::class)) {
+            DBALType::addType(GH9335IntObjectType::class, GH9335IntObjectType::class);
+        }
+
+        $this->setUpEntitySchema([GH9335Book::class, GH9335Author::class]);
+    }
+
+    /**
+     *  Verifies that entities with foreign keys with custom id object types don't throw an exception
+     *
+     * The test passes when refresh() does not throw an exception
+     */
+    public function testFlattenIdentifierWithObjectId(): void
+    {
+        $author = new GH9335Author('Douglas Adams');
+        $book   = new GH9335Book(new GH9335IntObject(42), 'The Hitchhiker\'s Guide to the Galaxy', $author);
+
+        $this->_em->persist($author);
+        $this->_em->persist($book);
+        $this->_em->flush($author);
+        $this->_em->flush($book);
+
+        $this->_em->refresh($book);
+
+        self::assertInstanceOf(GH9335IntObject::class, $book->getId());
+    }
+}
+
+
+class GH9335IntObjectType extends Type
+{
+    public function getSQLDeclaration(array $column, AbstractPlatform $platform): string
+    {
+        return $platform->getIntegerTypeDeclarationSQL($column);
+    }
+
+    public function getName(): string
+    {
+        return self::class;
+    }
+
+    public function convertToDatabaseValue($value, AbstractPlatform $platform): int
+    {
+        return $value->wrappedInt;
+    }
+
+    public function convertToPHPValue($value, AbstractPlatform $platform): GH9335IntObject
+    {
+        return new GH9335IntObject((int) $value);
+    }
+
+    public function getBindingType(): int
+    {
+        return ParameterType::INTEGER;
+    }
+
+    public function requiresSQLCommentHint(AbstractPlatform $platform): bool
+    {
+        return true;
+    }
+}
+
+class GH9335IntObject
+{
+    /** @var int */
+    public $wrappedInt;
+
+    public function __construct(int $wrappedInt)
+    {
+        $this->wrappedInt = $wrappedInt;
+    }
+
+    public function __toString(): string
+    {
+        return (string) $this->wrappedInt;
+    }
+}
+
+/**
+ * @Entity
+ */
+class GH9335Book
+{
+    /**
+     * @var GH9335IntObject
+     * @Id
+     * @Column(type=GH9335IntObjectType::class, unique=true)
+     */
+    protected $id;
```

Comment:
```suggestion
    private $id;
```

Please don't use `protected` if you don't need to. Please switch all properties to private.